[BUG] ensure proper constexpr of string hashing fix fails to build on 32-bit archs

The fix(strutil.h): ensure proper constexpr of string hashing (https://github.com/AcademySoftwareFoundation/OpenImageIO/pull/3901) commit fails to build on 32-bit archs.

https://github.com/AcademySoftwareFoundation/OpenImageIO/commit/3fd70cfd350d7dea1e1ebd5344e60fe0d6eda584

We noticed this with 2.4 but applies to 2.5 and newer. Reverting this commit alone and we're good building beyond 2.4.12 (upgrading from 2.4.11).

```
FAILED: src/libutil/CMakeFiles/strutil_test.dir/strutil_test.cpp.o
/pobj/openimageio-2.4.17.0/bin/c++ -DOIIO_INTERNAL=1 -DUSE_BOOST_FILESYSTEM -DUSE_EXTERNAL_PUGIXML=1 -DUSE_FREETYPE=1 -DUSE_JPEG_TURBO=1 -DUSE_OCIO=1 -DUSE_OPENCOLORIO=1 -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -I/pobj/openimageio-2.4.17.0/build-i386/include/OpenImageIO -I/pobj/openimageio-2.4.17.0/build-i386/include -I/pobj/openimageio-2.4.17.0/build-i386/src/include -I/pobj/openimageio-2.4.17.0/OpenImageIO-2.4.17.0/src/include -isystem /usr/local/include -isystem /usr/local/include/Imath -O2 -pipe -DNDEBUG -std=c++14 -Wall -Wno-unused-function -Wno-overloaded-virtual -Wno-unneeded-internal-declaration -Wno-unused-private-field -Wno-tautological-compare -Qunused-arguments -Wunknown-warning-option -Wno-unused-local-typedefs -Wno-expansion-to-defined -fno-math-errno -MD -MT src/libutil/CMakeFiles/strutil_test.dir/strutil_test.cpp.o -MF src/libutil/CMakeFiles/strutil_test.dir/strutil_test.cpp.o.d -o src/libutil/CMakeFiles/strutil_test.dir/strutil_test.cpp.o -c /pobj/openimageio-2.4.17.0/OpenImageIO-2.4.17.0/src/libutil/strutil_test.cpp
/pobj/openimageio-2.4.17.0/OpenImageIO-2.4.17.0/src/libutil/strutil_test.cpp:354:22: error: constexpr variable 'hash' must be initialized by a constant expression
    constexpr size_t hash = Strutil::strhash("much longer string");
                     ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/pobj/openimageio-2.4.17.0/OpenImageIO-2.4.17.0/src/include/OpenImageIO/detail/farmhash.h:1207:24: note: cannot refer to element -4 of array of 19 elements in a constant expression
  uint32_t a = Fetch(s - 4 + (len >> 1));
                       ^
/pobj/openimageio-2.4.17.0/OpenImageIO-2.4.17.0/src/include/OpenImageIO/detail/farmhash.h:1246:9: note: in call to 'Hash32Len13to24(&"much longer string"[0], 18, 0)'
        Hash32Len13to24(s, len);
        ^
/pobj/openimageio-2.4.17.0/OpenImageIO-2.4.17.0/src/include/OpenImageIO/detail/farmhash.h:2070:7: note: in call to 'Hash32(&"much longer string"[0], 18)'
      farmhashmk::Hash32(s, len));
      ^
/pobj/openimageio-2.4.17.0/OpenImageIO-2.4.17.0/src/include/OpenImageIO/detail/farmhash.h:2100:49: note: in call to 'Hash32(&"much longer string"[0], 18)'
  return sizeof(size_t) == 8 ? Hash64(s, len) : Hash32(s, len);
                                                ^
/pobj/openimageio-2.4.17.0/OpenImageIO-2.4.17.0/src/include/OpenImageIO/strutil.h:377:12: note: in call to 'Hash(&"much longer string"[0], 18)'
    return OIIO::farmhash::inlined::Hash(s, len);
           ^
/pobj/openimageio-2.4.17.0/OpenImageIO-2.4.17.0/src/include/OpenImageIO/strutil.h:389:25: note: in call to 'strhash(18, &"much longer string"[0])'
    return s.length() ? strhash(s.length(), s.data()) : 0;
                        ^
/pobj/openimageio-2.4.17.0/OpenImageIO-2.4.17.0/src/libutil/strutil_test.cpp:354:29: note: in call to 'strhash({&"much longer string"[0], 18})'
    constexpr size_t hash = Strutil::strhash("much longer string");
                            ^
1 error generated.
```

--- a/src/include/OpenImageIO/detail/farmhash.h
+++ b/src/include/OpenImageIO/detail/farmhash.h
@@ -2097,7 +2097,7 @@ STATIC_INLINE uint64_t Hash64(const char* s, size_t len) {
 // May change from time to time, may differ on different platforms, may differ
 // depending on NDEBUG.
 STATIC_INLINE size_t Hash(const char* s, size_t len) {
-  return sizeof(size_t) == 8 ? Hash64(s, len) : Hash32(s, len);
+  return sizeof(size_t) == 8 ? size_t(Hash64(s, len)) : size_t(Hash32(s, len));
 }
 
 // Hash function for a byte array.  For convenience, a 64-bit seed is also

--- a/src/include/OpenImageIO/strutil.h
+++ b/src/include/OpenImageIO/strutil.h
@@ -370,13 +370,13 @@ std::string OIIO_UTIL_API wordwrap (string_view src, int columns = 80,
 
 
 /// Our favorite "string" hash of a length of bytes. Currently, it is just
-/// a wrapper for an inlined, constexpr (if C++ >= 14), Cuda-safe farmhash.
+/// a wrapper for an inlined, constexpr, Cuda-safe farmhash.
 /// It returns a size_t, so will be a 64 bit hash on 64-bit platforms, but
 /// a 32 bit hash on 32-bit platforms.
 inline constexpr size_t
 strhash(size_t len, const char *s)
 {
-    return OIIO::farmhash::inlined::Hash(s, len);
+    return size_t(OIIO::farmhash::inlined::Hash64(s, len));
 }
 
 

--- a/src/libutil/strutil_test.cpp
+++ b/src/libutil/strutil_test.cpp
@@ -346,13 +346,13 @@ void
 test_hash()
 {
     using namespace Strutil;
-    OIIO_CHECK_EQUAL(strhash("foo"), 6150913649986995171);
-    OIIO_CHECK_EQUAL(strhash(std::string("foo")), 6150913649986995171);
-    OIIO_CHECK_EQUAL(strhash(string_view("foo")), 6150913649986995171);
+    OIIO_CHECK_EQUAL(strhash("foo"), size_t(6150913649986995171));
+    OIIO_CHECK_EQUAL(strhash(std::string("foo")), size_t(6150913649986995171));
+    OIIO_CHECK_EQUAL(strhash(string_view("foo")), size_t(6150913649986995171));
     OIIO_CHECK_EQUAL(strhash(""), 0);  // empty string hashes to 0
     // Check longer hash and ensure that it's really constexpr
     constexpr size_t hash = Strutil::strhash("much longer string");
-    OIIO_CHECK_EQUAL(hash, 16257490369375554819ULL);
+    OIIO_CHECK_EQUAL(hash, size_t(16257490369375554819ULL));
 }
 
 

